Split of the ITS threshold wf: Calibrator + aggregator - ADDED support for no EoS case

iravasen opened the pull request:
Changes included:
- o2-its-threshold-calib-workflow (already existing): foreseen to run on every EPN, ship DCSconfigObjects_t to aggregator in the endOfStream.
- o2-its-threshold-aggregator-workflow (new): foreseen to run on 1 EPN aggregator node, collects and merge all the DCSConfigObjects_t coming from every EPN. Prepare CCDB wrapper and payload and can ship this to ccdb-populator-wf (if endOfStream exists) or directly to CCDB with CCDBApi (i.e. no ccdb-populator-wf)
- fixes for ITHR tuning analysis

iravasen committed:
Split of ITS threshold calib workflow in two workflows: Calibrator (running of each EPN) + Aggregator (running on EPN aggregator)

iravasen committed:
o2-threshold-calib modified to work without end-of-stream

davidrohr reviewed (dismissed):
thx, will wait for the FullCI to pass and then merge for tomorrow's nightly.

shahor02 commented on Detectors/ITSMFT/ITS/workflow/src/ThresholdAggregatorSpec.cxx:
```diff
@@ -0,0 +1,220 @@
+  const auto tunString = pc.inputs().get<o2::dcs::DCSconfigObject_t>("tunestring");
+  std::string tmpString(tunString.begin(), tunString.end());
+  //Merge all strings coming from several sources (EPN)
+  std::copy(tmpString.begin(), tmpString.end(), std::back_inserter(tuningMerge));
```
Here you already create a` vector<char>`, then you create one more locally. Instead, you better do 
```
const auto tunString = pc.inputs().get<gsl::span<char>>("tunestring");
std::copy(tmpString.begin(), tmpString.end(), std::back_inserter(tuningMerge));
```
BTW, you don't need `this->` in the class methods.

shahor02 commented on Detectors/ITSMFT/ITS/workflow/src/ThresholdAggregatorSpec.cxx:
```diff
@@ -0,0 +1,220 @@
+
+  // Below is CCDB stuff
+  long tstart, tend;
+  tstart = o2::ccdb::getCurrentTimestamp();
+  constexpr long SECONDSPERYEAR = 365 * 24 * 60 * 60;
+  tend = o2::ccdb::getFutureTimestamp(SECONDSPERYEAR);
```
note that the `tstart = o2::ccdb::getCurrentTimestamp()` gives time in `ms`, while your `tend` is just 1 year in seconds, while it should be an abs. time-stamp in ms. If you want to assign 1y validity, you should have
```
tend = tstart + 365L*24*3600*1000;
```

shahor02 commented on Detectors/ITSMFT/ITS/workflow/src/ThresholdCalibratorSpec.cxx:
```diff
@@ -603,16 +609,16 @@ void ITSThresholdCalibrator::setRunType(const short int& runtype)
 {
   // In threshold scan case, reset mThresholdTree before writing to a new file
-  if ((this->mScanType == 'T') && ((this->mRowCounter)++ == N_ROWS_PER_FILE)) {
+  if ((!this->mCounter || this->mCheckEos) && (this->mScanType == 'T') && ((this->mRowCounter)++ == N_ROWS_PER_FILE)) {
```
Not sure I understand the logic: how is the data sent to the aggregator in absence of EoS?
I think the only way to do this is not to way for the finalize called from the stop() but detect that the scan for a given chip is complete and send its output from the regular `run()` call. This means, obviously, that each EPN should see more TFs than is needed to complete the scan for every chip it serves to.

shahor02 reviewed (changes requested):
sorry, pressed `approve` by mistake while I wanted `request changes`.

iravasen committed:
Improvements based on Ruben's comments + simplification of logic when operating without EoS

iravasen committed:
commented line of code removed

iravasen commented:
@shahor02 thanks a lot for the comments. I implemented the first 2 while for the third I removed those checks in the if condition trying to simplify a bit the logic when operating without EoS. The logic remains anyway a bit hard to understand but it is as follows:
- INFO: in our thr scan, when the scan reaches the last row of a chip the software behind (ITSComm) puts a zero in one of the calibration word user fields. The zeros come for the last 50 injections in the last row of every chipID. 
- We read the calibration word every time (for every ROF) in the run() method
- At the end the run() method, we check if the RunType (namely the calibration word user field mentioned above) is 0. We have to count the zeros because we expect 50 per chip. Then if runType is 0, we call finalize() which fill the DCSConfigObject_t only for the chips for which we have recorded enough data and then we ship out the DCSconfigObject with ```pc.outputs(...)```. 
- We do step above in a way that at the next RunType = 0, the data of older chips are cleared so that they do not end up again in the DCSconfigObject_t. This explains some calls like ```mThresholds.erase()``` or ```mTuning.clear()```.

davidrohr commented:
ok, I am not merging it tonight for the nightly. You said you need it on Tuesday. So we have to make sure it goes in tomorrow's nightly.

shahor02 reviewed (approved):
Hi @iravasen 
OK, thanks for clarifications, merging.
